Implementing transformation for Immunization

vsmahesh opened the pull request:
Transformation of Immunization data-type between Fhir and HealthVault.

There are a handful of fields in Fhir for which there are no specific fields present in HealthVault. They are ignored for the time being. Once we finalize an approach on using extension for maximum fidelity, we can revisit those fields (if required)

Merge Note that this is based on the procedure transformation branch.

Commit:
Adding transformation for Immunization to Fhir

1. Adding necessary unit tests
2. Adding necessary Extension constants and structure definition
3. Adding transformation for ApproximateDate to Fhir
4. Chaning method accessor to internal from public for
ApproximateTimeToFhirTime.cs
5. Adding transformation for Immunization to Fhir

Commit:
Formatting & Cleanup

1. Formatting ProcedureToFhir.cs
2. Cleaning up the way how TestCategory is populated

Commit:
Transformation for Immunization to HealthVault

Unit tests and associated resource for testing transformation of
Immunization to HealthVault
Transformation class for Immunization to HealthVault
Modification in FhirToHealthVaultDateTime in order to support
Hl7.Fhir.Model.Date
Bug fixes in Immunization to Fhir Transformation and associated test(s)

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,118 @@
+                hvImmunization.Consent = immunizationExtension.GetStringExtension(HealthVaultExtensions.ImmunizationDetailConcent);
+                hvImmunization.AdverseEvent = immunizationExtension.GetStringExtension(HealthVaultExtensions.ImmunizationDetailAdverseEvent);
+            }
+            if (!fhirImmunization.Note.IsNullOrEmpty())
```
nit: add extra line before

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,118 @@
+                    string sepearator = string.Empty;
+                    if (!string.IsNullOrEmpty(hvImmunization.CommonData.Note))
+                    {
+                        sepearator = "\n"; //Let's seperate each note with new line
```
Is this being used anywhere? Also, shouldn't we just use `Environment.NewLine` instead of a custom variable?

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,118 @@
+            {
+                fhirImmunization.Note.ForEach(note =>
+                {
+                    string sepearator = string.Empty;
```
Should be `separator`

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,118 @@
+                });
+            }
+
+
```
nit: remove extra line

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,118 @@
+            ItemTypes.Immunization hvImmunization = fhirImmunization.ToThingBase<ItemTypes.Immunization>();
+
+            hvImmunization.Name = fhirImmunization.VaccineCode.ToCodableValue();
+            hvImmunization.DateAdministrated = fhirImmunization.DateElement.ToAproximateDateTime();
```
Null check

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToFhir.cs:
```diff
@@ -72,6 +72,17 @@ internal static Hl7.Fhir.Model.Immunization ToFhirInternal(ItemTypes.Immunizatio
             fhirImmunization.Site = hvImmunization.AnatomicSurface.ToFhir();
+
+            fhirImmunization.Note.Add(new Hl7.Fhir.Model.Annotation {
+                Text = hvImmunization.CommonData.Note
```
Null check before adding.

vsmahesh replied:
This is not required since there is a built-in null check implemented in the library (Hl7.Fhir.Model) for most of the collections. If you look at it, the code for Note property is as the following:
 public List<Hl7.Fhir.Model.Annotation> Note
        {
            get { if(_Note==null) _Note = new List<Hl7.Fhir.Model.Annotation>(); return _Note; }
            set { _Note = value; OnPropertyChanged("Note"); }
        }

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ApproximateDateToFhir.cs:
```diff
@@ -0,0 +1,36 @@
+            return ApproximateDateToFhir.ToFhirInternal(approximateDate);
+        }
+    }
+    internal class ApproximateDateToFhir
```
nit: add extra line before.

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToFhir.cs:
```diff
@@ -0,0 +1,92 @@
+            return ImmunizationToFhir.ToFhirInternal(hvImmunization, hvImmunization.ToFhirInternal<Hl7.Fhir.Model.Immunization>());
+        }
+    }
+    internal static class ImmunizationToFhir
```
nit: add extra line before

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToFhir.cs:
```diff
@@ -0,0 +1,92 @@
+                {
+                    immunizationExtension.AddExtension(HealthVaultExtensions.ImmunizationDetailSequence, new FhirString(hvImmunization.Sequence));
+                }
+                if (!string.IsNullOrEmpty(hvImmunization.Consent))
```
nit: add extra line before

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToFhir.cs:
```diff
@@ -0,0 +1,92 @@
+                if (!string.IsNullOrEmpty(hvImmunization.Consent))
+                {
+                    immunizationExtension.AddExtension(HealthVaultExtensions.ImmunizationDetailConcent, new FhirString(hvImmunization.Consent));
+                }
```
nit: add extra line before

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToFhir.cs:
```diff
@@ -0,0 +1,92 @@
+            return fhirImmunization;
+        }
+
+        private static ResourceReference AddContainedResource(DomainResource domainResource, Resource resource, string id = null)
```
This feels more like a global helper than a private method. I believe we're doing very similar things already in procedure and the approach could be re-used.

brandonpollett reviewed (changes requested):
Now that procedure has been pulled in, can you rebase off of master and resolve the comments?

Commit:
Merge branch 'master' into immunization

Commit:
PR Comments

- Fixing typo
- Adding extra lines where ever required
- Adding null check for Date in ImmunizationToHealthVault.cs

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,122 @@
+                        separator = Environment.NewLine; //Let's seperate each note with new line
+                    }
+
+                    hvImmunization.CommonData.Note += $"{note.Text}";
```
I believe you want to use the separator here.

gcorvera commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,122 @@
+using Microsoft.HealthVault.ItemTypes;
+using System.Linq;
+using Hl7.Fhir.Model;
+using System;
```
Order the using statements

gcorvera commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,122 @@
+            }
+
+            if (string.IsNullOrEmpty(manufacturerComponent.Display))
+                return null;
```
open and close curly braces

gcorvera commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,122 @@
+            }
+
+            if (string.IsNullOrEmpty(practitionerComponent.Actor.Display))
+                return null;
```
open/close curly braces

gcorvera commented on Microsoft.HealthVault.Fhir/Transformers/ImmunizationToHealthVault.cs:
```diff
@@ -0,0 +1,122 @@
+                    return null;
+                }
+
+                return (containedReference as Practitioner).ToHealthVault();
```
>(containedReference as Practitioner [](start = 23, length = 35)

This could throw an exception, why not do the cast before checking for null?

vsmahesh replied:
It is done in the lambda expression to find containedReference

var containedReference = fhirImmunization.Contained.SingleOrDefault(resouce =>
                    resouce.Id.Equals(practitionerComponent.Actor.Reference) && resouce.GetType().Equals(typeof(Practitioner)));

gcorvera commented on Microsoft.HealthVault.Fhir.UnitTests/ToFhirTests/ImmunizationToFhirTests.cs:
```diff
@@ -0,0 +1,103 @@
+using NodaTime;
+using Hl7.Fhir.Model;
+using FhirOrganization = Hl7.Fhir.Model.Organization;
+using Microsoft.HealthVault.Fhir.Constants;
```
order

gcorvera reviewed (changes requested):
Please address the comments, this is close to be ready for merging.

Commit:
PR Comments

Notes are combined using new line separator
Sort & Removing usings in various files
Putting open-close braces where ever required

Commit:
Merge branch 'master' into immunization

Bringing latest changes into this PR
